### PR TITLE
(links): sanitation

### PR DESCRIPTION
**Pull request by rorychatt**
This pull request introduces comprehensive URL validation and sanitization across the codebase to prevent open redirect and injection vulnerabilities. The changes ensure that all URLs used in navigation, button creation, and client interactions are validated for safety. Additionally, new unit tests have been added to verify correct URL handling and to ensure that dangerous or malformed URLs are rejected or handled safely.

https://github.com/user-attachments/assets/15f2d3da-bae8-4d75-908b-8230e45e8b67

**Security Improvements:**

* Added URL validation to `IClientProvider.OpenUrl` and `IClientProvider.Redirect` extension methods, throwing an exception if the URL is invalid or unsafe. Also, restricts redirects to relative paths or same-origin URLs. (`Ivy/Client/ClientExtensions.cs`)
* Updated navigation logic to validate both external URLs and `appId` values before navigation, preventing open redirect and injection attacks. (`Ivy/Chrome/Shared.cs`) [[1]](diffhunk://#diff-af5110a30e34099b515e0d9bd6558fdcb966cda87b7c8abdf939955f5a167e2bL147-R163) [[2]](diffhunk://#diff-af5110a30e34099b515e0d9bd6558fdcb966cda87b7c8abdf939955f5a167e2bL82-R85)

**Testing and Validation:**

* Added `ButtonUrlValidationTests` to verify that the `Button.Url()` method properly validates, accepts, or rejects URLs, including edge cases and chaining. (`Ivy.Test/ButtonUrlValidationTests.cs`)
* Added `LinkBuilderTests` to ensure `LinkBuilder` correctly validates URLs, disables buttons for unsafe URLs, and handles various input scenarios. (`Ivy.Test/LinkBuilderTests.cs`)

**Codebase Consistency:**

* Ensured consistent use of the `Ivy.Utils` namespace for URL validation throughout the client extension methods. (`Ivy/Client/ClientExtensions.cs`)

### CODE REVIEW

**Comment by rorychatt** [comment]
@claude review this

File: Ivy.Samples.Shared/Apps/Tests/UrlValidationTestApp.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            | new Button("URL with Query & Fragment", variant: ButtonVariant.Link)
+                .Url("https://example.com/search?q=test&sort=date#results");
+
+        var invalidButtons = Layout.Vertical().Gap(8)
+            | SafeButtonWithUrl("JavaScript Protocol", "javascript:alert('XSS')")
+            | SafeButtonWithUrl("Data Protocol", "data:text/html,<script>alert('XSS')</script>")
+            | SafeButtonWithUrl("VBScript Protocol", "vbscript:msgbox('XSS')")
+            | SafeButtonWithUrl("File Protocol", "file:///etc/passwd")
+            | SafeButtonWithUrl("Malformed URL", "https://example.com:javascript:alert('XSS')")
+            | SafeButtonWithUrl("App Protocol with Query", "app://MyApp?param=value")
```
Comment (zachwolfe) [inline Ivy.Samples.Shared/Apps/Tests/UrlValidationTestApp.cs]: Why do app protocol URLs with queries need to be banned?

File: Ivy.Samples.Shared/Apps/Tests/UrlValidationTestApp.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            | SafeButtonWithUrl("Anchor with Colon", "#section:value")
```
Comment (zachwolfe) [inline Ivy.Samples.Shared/Apps/Tests/UrlValidationTestApp.cs]: Same here (anchors with colons)

**Review by zachwolfe — COMMENTED** [review]
I left a couple of comments but this mostly looks good to me.

I should also note the following about the "Url Validation Test" sample. Some of these points might be unrelated to the validation added in this PR:

- Clicking on the "App Protocol" link gives me this error (might be browser-specific): https://github.com/user-attachments/assets/e6428a29-f736-442e-8cb4-84ee1231d839
- In the valid markdown section, clicking on "Relative Path", "Relative with Query" or "App Protocols" does nothing and causes the server to print this to the console:
```
warn: Ivy.AppHub[0]
      Event 'OnLinkClick' for Widget 'dmfdbogc8k' not found.
```
- In the valid markdown section, clicking on "Anchor Link" just does nothing without printing.
- In the _invalid_ markdown section, clicking on any of the links does nothing (as noted above by @ArtemKhvorostianyi). Not sure if this is the intended behaviour for invalid links or not.

**Review by nielsbosma — APPROVED** [review]

**Comment by rorychatt** [comment]
I might have been too strict in what I banned.

Basically you could do `app://evil:javascript:alert('XSS')` to introduce a new protocol with multiple colons

Now you can do this

```js
// Good
app://MyApp?param=value
app://MyApp?param1=value1&param2=value2

// Banned
app://evil:javascript:alert('XSS')  // Multiple colons after app://
app://path#fragment                  // Fragments not allowed
```

I think we should start with this, and see what actual use cases we get and then broaden up the link url sanitation 

Same with anchor links

```js
// Good
#section:value
#anchor:colons

// Bannded
#section:javascript:alert('XSS')  // Contains ://
#anchor?query=value               // Query params not allowed

```

Also fixed the links not working in the md page.
